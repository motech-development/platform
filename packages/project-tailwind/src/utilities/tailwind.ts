import clsx from 'clsx';

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: string[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: string[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: string[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: string[] = [];

    if (classNames.classNames) {
      finalStyles = [...finalStyles, ...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  return {
    createStyles,
  };
}
