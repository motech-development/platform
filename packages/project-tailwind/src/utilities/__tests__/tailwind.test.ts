import { useTailwind } from '../tailwind';

describe('tailwind', () => {
  it('should return the correct styles', () => {
    const { createStyles } = useTailwind('primary', 'md');

    expect(
      createStyles({
        classNames: ['text-sm', 'font-medium'],
        sizing: {
          md: ['md-4'],
          none: ['md-0'],
        },
        theme: {
          primary: ['bg-blue-50'],
          warning: ['bg-yellow-50'],
        },
      }),
    ).toMatchSnapshot();
  });

  it('should throw an error if theme is not defined', () => {
    const { createStyles } = useTailwind('danger', 'md');

    expect(() => {
      createStyles({
        classNames: ['text-sm', 'font-medium'],
        sizing: {
          md: ['md-4'],
          none: ['md-0'],
        },
        theme: {
          primary: ['bg-blue-50'],
          warning: ['bg-yellow-50'],
        },
      });
    }).toThrow('No classes defined for this theme.');
  });

  it('should throw an error if sizing is not defined', () => {
    const { createStyles } = useTailwind('primary', 'sm');

    expect(() => {
      createStyles({
        classNames: ['text-sm', 'font-medium'],
        sizing: {
          md: ['md-4'],
          none: ['md-0'],
        },
        theme: {
          primary: ['bg-blue-50'],
          warning: ['bg-yellow-50'],
        },
      });
    }).toThrow('No classes defined for this sizing.');
  });

  it('should omit universal classes if none are passed', () => {
    const { createStyles } = useTailwind('primary', 'md');

    expect(
      createStyles({
        sizing: {
          md: ['md-4'],
          none: ['md-0'],
        },
        theme: {
          primary: ['bg-blue-50'],
          warning: ['bg-yellow-50'],
        },
      }),
    ).toMatchSnapshot();
  });

  it('should omit theme styles if none are passed', () => {
    const { createStyles } = useTailwind('primary', 'md');

    expect(
      createStyles({
        classNames: ['text-sm', 'font-medium'],
        sizing: {
          md: ['md-4'],
          none: ['md-0'],
        },
      }),
    ).toMatchSnapshot();
  });

  it('should omit sizing styles if none are passed', () => {
    const { createStyles } = useTailwind('primary', 'md');

    expect(
      createStyles({
        classNames: ['text-sm', 'font-medium'],
        theme: {
          primary: ['bg-blue-50'],
          warning: ['bg-yellow-50'],
        },
      }),
    ).toMatchSnapshot();
  });
});
