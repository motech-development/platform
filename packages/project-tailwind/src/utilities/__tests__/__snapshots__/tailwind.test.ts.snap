// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwind should omit sizing styles if none are passed 1`] = `"text-sm font-medium bg-blue-50"`;

exports[`tailwind should omit theme styles if none are passed 1`] = `"text-sm font-medium md-4"`;

exports[`tailwind should omit universal classes if none are passed 1`] = `"md-4 bg-blue-50"`;

exports[`tailwind should return the correct styles 1`] = `"text-sm font-medium md-4 bg-blue-50"`;
